File: backend/tests/apply_fixes.py
#!/usr/bin/env python3
"""Apply all API.php fixes that the editor couldn't match due to quote escaping."""
import sys

f = open('backend/api.php', 'r', encoding='utf-8')
content = f.read()
f.close()

# ─── Fix 5: Duplicate prevention + name length validation in studentSaveCard ───
# Use a smaller, unique anchor that will match
old_block = """    $studentName=trim((string)($data['student_name']??''));
    $idType=trim((string)($data['id_type']??''));
    if ($studentName==='') response(['success'=>false,'message'=>'Please enter the student full name'],422);
    if (!in_array($idType,['COLLEGE','JUNIOR_HIGH','SENIOR_HIGH'],true)) response(['success'=>false,'message'=>'Please select a valid department'],422);
    $fields=['student_name','id_type','course','grade_level','section_name','student_number','student_id_number','lrn','academic_year','school_year','photo_path','address_line1','address_line2','emergency_label','emergency_contact','emergency_phone','terms_title','term_1','term_2','term_3','institution_name','institution_address','mobile_no','telephone_no','email_address','signatory_id','signatory_name','signature_path','template_id'];
    $values=[];foreach($fields as $f)$values[$f]=trim((string)($data[$f]??''));
    $cols=implode(',',$fields);$pars=implode(',',array_map(fn($f)=>\":$f\",$fields));
    $stmt=$pdo->prepare(\"INSERT INTO id_cards($cols) VALUES($pars)\");$stmt->execute($values);
    $id=(int)$pdo->lastInsertId();
    response(['success'=>true,'id'=>$id,'message'=>'Your ID request has been submitted successfully.']);
  }
  if ($action==='lostIdRequest'

File: backend/tests/apply_fixes.php
<?php
// Fix script: applies duplicate prevention + length validation to studentSaveCard
// Uses line-number based insertion since we know the exact line numbers
$file = __DIR__ . '/../api.php';
$lines = file($file, FILE_IGNORE_NEW_LINES);
$newLines = [];

foreach ($lines as $i => $line) {
    $newLines[] = $line;
    
    // Line 134 (0-indexed 133) is: $idType=trim((string)($data['id_type']??''));
    // After it, add validation code
    if ($i === 133) {
        $newLines[] = '    /* Input length validation */';
        $newLines[] = "    \$course=trim((string)(\$data['course']??''));";
        $newLines[] = "    if (\$studentName !== '' && strlen(\$studentName) > 150) response(['success'=>false,'message'=>'Student name is too long (max 150 characters)'],422);";
        $newLines[] = "    if (strlen(\$idType) > 20) response(['success'=>false,'message'=>'Invalid department'],422);";
        $newLines[] = "    if (strlen(\$course) > 255) response(['success'=>false,'message'=>'Course is too long'],422);";
    }
    
    // Line 138 (0-indexed 137) is: $values=[];foreach($fields as $f)$values[$f]=trim((string)($data[$f]??''));
    // After it, add duplicate prevention
    if ($i === 137) {
        $newLines[] = '    /* Duplicate prevention */';
        $newLines[] = "    \$sn=trim((string)\$values['student_number']);";
        $newLines[] = "    \$sid=trim((string)\$values['student_id_number']);";
        $newLines[] = "    \$lrn=trim((string)\$values['lrn']);";
        $newLines[] = "    \$dupCols=[];\$dupParams=[];";
        $newLines[] = "    if (\$sn !== '')  { \$dupCols[]='student_number=?';      \$dupParams[]=\$sn; }";
        $newLines[] = "    if (\$sid !== '') { \$dupCols[]='student_id_number=?';   \$dupParams[]=\$sid; }";
        $newLines[] = "    if (\$lrn !== '') { \$dupCols[]='lrn=?';                 \$dupParams[]=\$lrn; }";
        $newLines[] = "    if (\$dupCols) {";
        $newLines[] = "      \$dupParams[]=\$studentName;";
        $newLines[] = "      \$dupCheck=\$pdo->prepare('SELECT id FROM id_cards WHERE student_name=? AND ('.implode(' OR ',\$dupCols).') LIMIT 1');";
        $newLines[] = "      \$dupCheck->execute(\$dupParams);";
        $newLines[] = "      \$dup = \$dupCheck->fetch();";
        $newLines[] = "      if (\$dup) response(['success'=>false,'message'=>'A student with this name and student number/ID/LRN already exists.',409]);";
        $newLines[] = "    }";
    }
}

file_put_contents($file, implode("\n", $newLines) . "\n");
echo "Fixes applied.\n";

